src/protocol/sort/shuffle.rs

```diff
@@ -156,21 +158,21 @@ impl<'a, F: Field> Shuffle<'a, F> {
         &mut self,
         ctx: ProtocolContext<'_, Replicated<F>, F>,
         permutations: &mut (Permutation, Permutation),
-    ) -> Result<(), BoxError>
+    ) -> Result<Vec<Replicated<F>>, BoxError>
     where
         F: Field,
     {
-        *self.input = self
+        self.input = self
             .shuffle_or_unshuffle_once(ShuffleOrUnshuffle::Shuffle, &ctx, Step1, permutations)
             .await?;
-        *self.input = self
+        self.input = self
             .shuffle_or_unshuffle_once(ShuffleOrUnshuffle::Shuffle, &ctx, Step2, permutations)
             .await?;
-        *self.input = self
+        self.input = self
             .shuffle_or_unshuffle_once(ShuffleOrUnshuffle::Shuffle, &ctx, Step3, permutations)
             .await?;
 
```

Comment:
given you will immediately `take` this input on the next line, maybe just skip setting it to `self`?
```suggestion
        let input = self
            .shuffle_or_unshuffle_once(ShuffleOrUnshuffle::Shuffle, &ctx, Step3, permutations)
            .await?;
        Ok(input)
```
or simply
```suggestion
        Ok(self
            .shuffle_or_unshuffle_once(ShuffleOrUnshuffle::Shuffle, &ctx, Step3, permutations)
            .await?)
```